### PR TITLE
Check user for KV http requests

### PR DESCRIPTION
**Pull request by mberhault**
Verify certificate user vs request user in KV http endpoints.
TS remains "check certificate settings only" since it has no associated
user. All other http endpoints remain the same.

Refactor tests a bit, and add http authentication for kv http endpoint.

### CODE REVIEW

**Comment by vivekmenezes** [comment]
I suppose I will need to do the same for the sqlserver right?

**Comment by MBerhault** [comment]
Yup. The RPC endpoint should be taken care of already, although it would be good to add a test for it. The HTTP one will be similar to this one (eg: within the http handler, building the hook before decoding, and applying the hook function after decoding the proto)

**Comment by vivekmenezes** [comment]
Okay will do once you're checked this in.

File: server/server.go [hunk 1]
```diff
@@ -196,7 +195,11 @@ func (s *Server) initHTTP() {
 	s.mux.Handle(debugEndpoint, s.admin)
 	s.mux.Handle(statusKeyPrefix, s.status)
 
-	s.mux.HandleFunc(kv.DBPrefix, s.authenticateRequest(s.kvDB))
+	// KV handles its own authentication, verifying user certificates against
+	// the requested user.
+	s.mux.Handle(kv.DBPrefix, s.kvDB)
+	// TS requests do not have a user, so only check that client certificates are
```
Comment (spencerkimball) [inline server/server.go]: TS requests certainly should have a user, so we're going to want a TODO here.

File: server/authentication_test.go [hunk 2]
```diff
@@ -42,83 +43,70 @@ func TestSSLEnforcement(t *testing.T) {
...
 	testCases := []struct {
-		method, key   string
-		certsStatus   int // Status code for https with client certs.
-		noCertsStatus int // Status code for https without client certs.
+		method, key string
+		ctx         *base.Context
+		success     bool // request sent successfully (may be non-200)
+		code        int  // http response code
 	}{
 		// /ui/: basic file server: no auth.
```
Comment (spencerkimball) [inline server/authentication_test.go]: Fix comments on all lines about "no auth".

**Comment by spencerkimball** [comment]
LGTM

File: server/server.go (hunk 1 above)
Comment (MBerhault) [inline server/server.go]: Added todo in comment.

File: server/authentication_test.go (hunk 2 above)
Comment (MBerhault) [inline server/authentication_test.go]: no, that's correct, only the KV requests have authentication applied.
the insecureContext attempts to speak HTTP to an HTTPS server, which fails. however, client certificates are not checked, hence the "no auth" comment.

**Comment by MBerhault** [comment]
Uh. Vivek, I don't think you'll be able to do more than what's already in https://github.com/cockroachdb/cockroach/pull/1476 since there's no user in the wire format.

**Comment by vivekmenezes** [comment]
Got it. Thanks

**Comment by MBerhault** [comment]
@MattTracy: see TODO in server.go: there's no header in the timeseries requests, so we don't have a requested user. Was it just left out, or is there a reason not to have one?

**Comment by bdarnell** [comment]
LGTM

**Comment by mrtracy** [comment]
@MBerhault: It was just left out, I don't think any of the status endpoints have a user right now.

**Comment by MBerhault** [comment]
added TODO for the new sql wire endpoint as well.

**Pull request merged**